src/providers/idp/idp_auth.c

```diff
@@ -0,0 +1,543 @@
+    if (extra_args[c] == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "talloc_asprintf failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+    c++;
+
+    extra_args[c] = talloc_asprintf(extra_args,
+                                    "--userinfo-endpoint=%s",
+                                    idp_auth_ctx->userinfo_endpoint);
+    if (extra_args[c] == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "talloc_asprintf failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+    c++;
+
+    if (idp_auth_ctx->idp_type != NULL
+            && strncasecmp(idp_auth_ctx->idp_type, "keycloak:", 9) == 0) {
+        /* Keycloak is using the 'id' attribute as 'sub' for OIDC */
+        extra_args[c] = talloc_asprintf(extra_args,
+                                        "--user-identifier-attribute=sub");
+    } else {
+        extra_args[c] = talloc_asprintf(extra_args,
+                                        "--user-identifier-attribute=id");
+    }
+    if (extra_args[c] == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "talloc_asprintf failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+    c++;
+
+    extra_args[c] = talloc_asprintf(extra_args,
+                                    "--scope=%s",
+                                    idp_auth_ctx->scope);
+    if (extra_args[c] == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "talloc_asprintf failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+    c++;
+
+    chain_id = sss_chain_id_get();
+    extra_args[c] = talloc_asprintf(extra_args,
+                                    "--chain-id=%lu",
+                                    chain_id);
+    if (extra_args[c] == NULL) {
+        DEBUG(SSSDBG_OP_FAILURE, "talloc_asprintf failed.\n");
+        ret = ENOMEM;
+        goto done;
+    }
+    c++;
+
+    if (DEBUG_IS_SET(SSSDBG_TRACE_LIBS)) {
+        extra_args[c] = talloc_asprintf(extra_args, "--libcurl-debug");
+        if (extra_args[c] == NULL) {
+            DEBUG(SSSDBG_OP_FAILURE, "talloc_asprintf failed.\n");
+            ret = ENOMEM;
+            goto done;
+        }
+        c++;
+    }
+
+    extra_args[c] = NULL;
+
+    *oidc_child_extra_args = extra_args;
+
+    ret = EOK;
+
+done:
+
+    if (ret != EOK) {
+        talloc_free(extra_args);
+    }
+
+    return ret;
+}
+
+static const char *get_stored_request_data(TALLOC_CTX *mem_ctx,
+                                           struct idp_auth_ctx *idp_auth_ctx,
+                                           struct pam_data *pd)
+{
+    int ret;
+    const char *send_data;
+    struct idp_open_req_data *open_req = NULL;
+    const char *user_code;
+    size_t user_code_len;
+
+    if (sss_authtok_get_type(pd->authtok) != SSS_AUTHTOK_TYPE_OAUTH2) {
+        DEBUG(SSSDBG_OP_FAILURE,
+              "Unexpected authentication type [%d][%s].\n",
+              sss_authtok_get_type(pd->authtok),
+              sss_authtok_type_to_str(sss_authtok_get_type(pd->authtok)));
+        goto done;
```

Comment:
`send_data` will be returned uninitialized here.

Reply:
And below.

Reply:
Fixed